Source/DafnyLanguageServer/DafnyLanguageServer.cs

```diff
@@ -19,21 +20,46 @@ public static class DafnyLanguageServer {
       }
     }
 
-    public static LanguageServerOptions WithDafnyLanguageServer(this LanguageServerOptions options, IConfiguration configuration) {
+    public static LanguageServerOptions WithDafnyLanguageServer(this LanguageServerOptions options,
+      IConfiguration configuration, CancellationTokenSource cancelLanguageServer) {
       return options
         .WithDafnyLanguage(configuration)
         .WithDafnyWorkspace(configuration)
         .WithDafnyHandlers()
-        .OnInitialize(InitializeAsync)
+        .OnInitialize((server, @params, token) => InitializeAsync(server, @params, token, cancelLanguageServer))
         .OnStarted(StartedAsync);
     }
 
-    private static Task InitializeAsync(ILanguageServer server, InitializeParams request, CancellationToken cancellationToken) {
+    private static Task InitializeAsync(ILanguageServer server, InitializeParams request, CancellationToken cancelRequestToken,
+        CancellationTokenSource cancelLanguageServer) {
       var logger = server.GetRequiredService<ILogger<Program>>();
       logger.LogTrace("initializing service");
+
+      KillLanguageServerIfParentDies(request, cancelLanguageServer);
+
       return Task.CompletedTask;
     }
 
+    /// <summary>
+    /// As part of the LSP spec, a language server must kill itself if its parent process dies
+    /// https://github.com/microsoft/language-server-protocol/blob/gh-pages/_specifications/specification-3-16.md?plain=1#L1713
+    /// </summary>
+    private static void KillLanguageServerIfParentDies(InitializeParams request, CancellationTokenSource cancelLanguageServer) {
+      if (!(request.ProcessId >= 0)) {
+        return;
+      }
+
+      try {
+        var hostProcess = Process.GetProcessById((int)request.ProcessId)!;
+        hostProcess.EnableRaisingEvents = true;
+        hostProcess.Exited += (_, _) => cancelLanguageServer.Cancel();
+      }
+      catch {
+        // If the process dies before we get here then request shutdown immediately
+        cancelLanguageServer.Cancel();
```

Comment:
This will make it harder to test the server independently, because it will prevent us from replaying previously-recorded LSP traces.  Can we find a trick to have the server ignore this in some cases?  Maybe it's enough to do nothing in this catch case?

Reply:
Would it be possible to change the ParentId in the Initialize request when replaying the messages?

Reply:
Yes, certainly.